### PR TITLE
[WIP] Add support for Sailfish OS

### PR DESCRIPTION
**Pull request by Max-Might**
Hello.

I have been working on porting the engine to the [Sailfish](https://sailfishos.org/) operating system.
I decided to open this PR to track progress and invite other interested parties to join the effort.

Sailfish is a Linux-based operating system for mobile devices.
It uses a lot of the common platform software stack like D-Bus, systemd, PulseAudio, Qt and others.
It also uses Wayland instead of x server.

This is a work in progress. Here is what works at the moment.

- [x] Rendering. The port uses QOpenGlWindow for accelerated drawing.
- [x] Audio. Reusing the PA driver we already have. Also uses `libaudioresource` to manage permissions.
- [x] Mouse and keyboard input.
- [ ] Touch events are not sent (easy).
- [ ] Landscape mode. Investigating how to implement landscape mode. May need to rotate canvas and translate input.
- [ ] Exporter. SailfishOS uses RPM package format. My idea is to export the project files to a directory with a `.spec` file and call `spectacle` to generate the RPM. This way we will not introduce any additional dependencies like `librpm` nor resort to manually generating the RPM.

### CODE REVIEW

File: platform/sailfish/os_sailfish.cpp [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+
+	return false;
+}
+
+void OS_Sailfish::set_cursor_shape(CursorShape p_shape) {
+	// TODO: implement
+}
+
+Point2 OS_Sailfish::get_mouse_position() const {
+	// DODO: implement
```
Comment (karroffel) [inline platform/sailfish/os_sailfish.cpp]: Didn't those went extinct a while ago?

File: platform/sailfish/godot_sailfish.cpp [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+	Error error = Main::setup(argv[0], argc-1, &argv[1]);
+	if (error != OK) {
+		return 255;
+	}
+
+	if (Main::start()) {
+		os.run(); // it is actually the OS that decides how to run
+	}
+
+	delete os.window;
```
Comment (karroffel) [inline platform/sailfish/godot_sailfish.cpp]: Shouldn't this be `memdelete()` and `memnew()` a few lines above? 😛

**Comment by punto-** [comment]
Is the export process portable? can it run on a windows/osx pc? What phones
run this? is the "QOpenGlWindow" a Sailfish specific thing, or a Wayland
specific thing?

**Comment by karroffel** [comment]
This looks really awesome!! Too bad I can't test, but I'd love if Godot would support sailfish!

Also, you said that it's using Wayland, couldn't some part of this be split into a OS_Wayland or something so it could be used for "regular" Linux systems too?

File: platform/sailfish/godot_sailfish.cpp (hunk 2 above)
Comment (punto-) [inline platform/sailfish/godot_sailfish.cpp]: no this is fine, memnew doesn't exist before OS is initialized, so those things have to be allocated the old fashion way

**Comment by karroffel** [comment]
@punto- Sailfish seems to use Qt, so this `QOpenGLWindow` is Qt specific, not wayland

**Comment by punto-** [comment]
yeah.. the "Q" was suspicious.. (I was also thinking this could be reused for Wayland). I wonder if it's possible to avoid Qt altogether, since you only need it for the window and gl context management, how much smaller would the binary be?

**Comment by karroffel** [comment]
Looks like Sailfish apps need to be in Qt, that's why there are qml files and all

**Comment by punto-** [comment]
they don't *need* to be in Qt, just like Android doesn't *need* to be in java, the question is, do they *force* you to be in Qt? There's a window server and a kernel underneath Qt, can't we just access that directly?

**Comment by Max-Might** [comment]
The Qt framework is the official API for developing Sailfish apps.
I went for it as the easiest way to obtain an OpenGL window.
Removing the dependency will not reduce executable by much because it is dynamically linked.

AFAIK the SDL2 port for sailfish uses the Wayland API instead of Qt but I do not know much about it.
If OS_Wayland materializes and works for this use case we can always update the port.
Also I am not sure how wayland handles peripherals. For desktop it looks like everybody is going for `libinput`.

As for the exporter, basically the Sailfish SDK is a Virtualbox VM with cross-compilers for all supported platforms (arm and i486 at the moment, aarch64 in progress).

**Comment by akien-mga** [comment]
Ah that makes sense, at first I was thinking "awesome, Wayland support for Linux", but I now understand that you did not have to implement direct Wayland support but get it directly from Qt 5. I don't think the dependency on Qt is too big a problem if it's only for SailfishOS, where it will indeed be dynamically linked.

**Comment by raymoo** [comment]
Maybe Wayland support here could be used for desktop Wayland support? I'm particularly interested in desktop multitouch support.

**Comment by akien-mga** [comment]
> Maybe Wayland support here could be used for desktop Wayland support? I'm particularly interested in desktop multitouch support.

Read my comment right above yours.

**Comment by raymoo** [comment]
Oh, I see. But this still could potentially be used on desktop platforms, right?

**Comment by akien-mga** [comment]
> Oh, I see. But this still could potentially be used on desktop platforms, right?

No, because we don't want to depend on huge Qt on desktop just for the sake of having Wayland support. On SailfishOS it makes sense as it's a core component of the ecosystem, but not on general Linux desktop.

**Comment by punto-** [comment]
For compiling they have a cross platform SDK, but I meant exporting the
game, and deploying to a phone during development, does the SDK have tools
to create the RPMs on windows and osx? Can people get this tool separately?
(so they don't need the entire sdk if they're not compiling the engine,
only using it with the export templates)

Yeah I was thinking of reusing Wayland code but I guess Qt is not that
bad.. I assume SDL has peripheral support so it's possible to use them with
Wayland

**Comment by Max-Might** [comment]
@punto- The SDK consists of two VMs - one for development and one for the emulator.
All the tools needed for build a package are in the development VM. The SDK also contains a package validator so you can check if your RPMs are compatible with Harbour (the application store).

Even when you are building a normal Sailfish application you do not need to run anything on your host besides the IDE and the VMs. When you hit 'Run' the IDE transfers your project files to the dev VM, compiles them, creates a package and deploys it to the emulator VM or real device. I think the same workflow would be possible with the exporter. The other option is just to export the files somewhere on the host and them manually SSH to the dev vm and run the packaging commands.

The [SailfishOS SDK](https://sailfishos.org/wiki/Application_SDK) is available for Linux 32/64, OS X and Windows.

**Comment by punto-** [comment]
Right, so what I'm saying is, can we avoid the dev VM for someone who is
only exporting a game from the editor? create the RMP and deploy it to the
device or emulator? a VM sounds like a huge requirement

**Comment by Max-Might** [comment]
To create an RPM package we have a few options:

1. Use the RPM packaging tools on the host.
On linux some distros have RPM-related tools in their repos. I don't think they are available for windows and OS X.

2. Use the RPM packaging tools in the SDK VM. This would be nice because the SDK is available for all major operating systems and you will also be able to use the RPM validator. And you will probably need the emulator VM for development or testing anyway.

3. Use `librpm` to build the package but this will introduce a new hard dependency to the engine since it has to be available for all platforms. I do not think this is a good idea.

Building the package 'manually' (without the tools that are intended to do just that) is not a good idea in my opinion since this is error-prone and not future-proof.

If anyone has any other ideas about this please feel free to join the discussion.
Anyway, before working on the exporter I would like to resolve some of the other issues (like the landscape mode).

**Comment by akien-mga** [comment]
The 4th option is to implement the RPM file format specification to be able to write it, but that's basically rewriting part of librpm: http://ftp.rpm.org/max-rpm/s1-rpm-file-format-rpm-file-format.html

Since Sailfish OS is a relatively "exotic" platform, I think it's not too much to ask of developers who want to develop to this platform to get the upstream SDK.

**Comment by punto-** [comment]
Sure, I'm not saying these are requirements to add the platform (the ios
export is not exactly ideal and it's a pretty popular platform :p ), only
that it'll make it more likely that people will export to it

**Comment by punto-** [comment]
In fact I bet we could ship a skeleton rpm as the export template, during export add the game data and update the manifest, etc, and re-zip, like we do with apks, no need to require the whole dev vm..                                                                                                                                                                                                                                                                                                                                        Sent from my BlackBerry 10 smartphone.                                                                                                                                                                                                                From: Ariel ManzurSent: Thursday, June 29, 2017 12:17 PMTo: godotengine/godotCc: godotengine/godot; MentionSubject: Re: [godotengine/godot] [WIP] Add support for Sailfish OS (#9435)Sure, I'm not saying these are requirements to add the platform (the ios export is not exactly ideal and it's a pretty popular platform :p ), only that it'll make it more likely that people will export to itOn 29 June 2017 at 11:45, Rémi Verschelde <notifications@github.com> wrote:The 4th option is to implement the RPM file format specification to be able to write it, but that's basically rewriting part of librpm: http://ftp.rpm.org/max-rpm/s1-rpm-file-format-rpm-file-format.html
Since Sailfish OS is a relatively "exotic" platform, I think it's not too much to ask of developers who want to develop to this platform to get the upstream SDK.

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

**Comment by Max-Might** [comment]
I am looking for a way to add landscape support but it looks like Qt does not have any APIs for this.
It can only let the compositor know you are already drawing in landscape.

@reduz Do you know if it is possible to use GLES on the OS class level to rotate the screen? I guess it would be possible with GLES? I have absolutely no experience with low level graphics :(

I know about glRotate() but it is not available in GLES 3?

**Comment by savegame** [comment]
i build that )) its godots editor in sailfish )) fun and nice )

**Comment by savegame** [comment]
Ok, 2D project ( simple test with 3 animated sprites ) is run fine! ) , but running 3D is segfault ( gles3 shaders cant compile )

**Comment by akien-mga** [comment]
Hi there, what's the current status on this PR?

**Comment by savegame** [comment]
@akien-mga , i think need just make that issue #576, and godot automaticly willwork on sailfish =) 
i am already porting [irricht engine](https://together.jolla.com/question/165688/nearly-done-irrlicht-19-opengles2-engine-port/) to sailfish, and i think in that way we can port godot, but now a dont have much time for this, maybe later
ps sorry for my english

**Comment by Max-Might** [comment]
@akien-mga My question still stands.
In order to support landscape mode we need to modify the view port from the OS class.
The input can be translated by some Qt helpers but I do not know how to do the OpenGL viewport rotation.

That is why I asked if it is possible/suitable to do it from the OS class.

@savegame This is nice but a generic Wayland port will not be enough here.
To port Godot to a new platform you also need sound, input(kbd,mouse,joystick), packaging and deployment tools etc. For example this port uses libaudioresource to request permission to play sound by the pulseaudio driver. Whether the port uses Wayland directly or Qt is not a big difference.

**Comment by savegame** [comment]
@Max-Might, in wayland no problem to get keyboard, mouse and touchscreen. But i don't now about sound, i am not test it yet.

**Comment by aaronfranke** [comment]
@Max-Might Are you still working on this?

I noticed that you mentioned GLES. How does Godot's decision to add Vulkan support affect this PR?

**Comment by akien-mga** [comment]
Closing as it stalled.

Support for SailfishOS would still be interesting to have, and any interested contributor could start from the work in this branch. As it's a relatively niche platform we'll need at least two things for it to be viable:
- a mostly feature-complete initial implementation
- an active maintainer to keep it up to date with changes in the master branch

**Comment by savegame** [comment]
@Max-Might, if you still interesting with Sailfish platform, i am allready build both Godot 2  and 3 for sailfish OS (i am get sample of use  audioresource from your code, thanks), and in Godot 3 i found a way to rotate render for Sailfish landscape orientation (just fix canvas shader in drivers gles2 ).

**Pull request closed without merging**